### DISCUSSION TITLE
How to append to List?

### DISCUSSION THREAD

**Discussion by refaelsh** [jtdaugherty/brick#470] [Q&A] [answered]
Hi,

There is no append function to List. 
1. Why? I was surprised there is no such a function. Seems natural there should be such a function.
2. How can I append to List? Please help :-)

**Comment by jtdaugherty** [comment]
1. Because a `List` is not intended to be used as a mutable data structure with all of the usual "list" operations; it is a UI representation of a list, and the state changes in the list should be deliberate. But this shouldn't be a problem, because ...
2. Your best bet is likely to use [listReplace](https://hackage.haskell.org/package/brick-1.9/docs/Brick-Widgets-List.html#v:listReplace) to provide the new element sequence you want to store in the list. If you want to get the elements from the list and then modify that sequence in some way (such as to put it back with `listReplace`), you can use [listElements](https://hackage.haskell.org/package/brick-1.9/docs/Brick-Widgets-List.html#v:listElements).

**Comment by refaelsh** [comment]
> Because a List is not intended to be used as a mutable data structure with all of the usual "list" operations; it is a UI representation of a list, and the state changes in the list should be deliberate. But this shouldn't be a problem, because ...

I understand.

> Your best bet is likely to use [listReplace](https://hackage.haskell.org/package/brick-1.9/docs/Brick-Widgets-List.html#v:listReplace) to provide the new element sequence you want to store in the list. If you want to get the elements from the list and then modify that sequence in some way (such as to put it back with listReplace), you can use [listElements](https://hackage.haskell.org/package/brick-1.9/docs/Brick-Widgets-List.html#v:listElements).

I dont get it. How `listReplace` helps me to append? Can you please provide a small example?

**Reply by jtdaugherty** [reply, marked as answer]
The function to modify the list would be something like `\lst -> listReplace (listElements lst <> elementsToAppend) (Just i) lst` where `i` is the index of the element that you want selected after the modification. In the context of `EventM`, you'd either use this with a lens (i.e. assuming `appStateList` is a lens made from the `_appStateLens` field of your application state type) like
```haskell
lst <- use appStateList
let newElements = listElements lst <> elementsToAppend
appStateList %= listReplace newElements (Just i)
```
or without a lens (i.e. assuming `appStateList` is just a record field of your application state type) like
```haskell
lst <- gets appStateList
let newElements = listElements lst <> elementsToAppend
modify $ \s -> s { appStateList = listReplace newElements (Just i) lst }
```

**Reply by refaelsh** [reply]
I understand. Thank you very much.

**Reply by jtdaugherty** [reply]
Okay, great, I'm glad that is helpful!
